src/components/tab-bar/tab-bar.tsx

```diff
@@ -175,12 +175,25 @@ export class TabBar {
         const oldSelectedTab = this.tabs.find((tab) => tab.active === true);
 
         if (oldSelectedTab) {
-            this.changeTab.emit({ ...oldSelectedTab, active: false });
+            this.replaceTab({ ...oldSelectedTab, active: false });
         }
 
-        this.changeTab.emit({ ...this.tabs[index], active: true });
+        this.replaceTab({ ...this.tabs[index], active: true });
     }
 
+    private replaceTab(tab: Tab) {
+        this.tabs = this.tabs.map(this.mapTab(tab));
+        this.changeTab.emit(tab);
+    }
+
+    private mapTab = (newTab: Tab) => (oldTab: Tab) => {
+        if (newTab.id === oldTab.id) {
+            return newTab;
+        }
+
+        return oldTab;
+    };
```

Comment:
🤯

Ok, so after a few minutes of trying to parse this code, what I think it's basically doing is setting the `active` property of the previously selected tab to `false` and that of the newly selected tab to `true`, and it also makes sure to recreate the whole array, in order to trigger a re-render…

Am I somewhat close?

Reply:
I don't think this by itself will trigger a re-render, we just need to make sure that our saved tabs (`this.tabs`) are updated in case a new tab is clicked before `this.tabs` have had a chance to be updated by the consumer. I think that the consumer update is synchronous (i.e. the consumer _thinks_ that they have updated the props before a new tab is selected), but it seems that Stencil queues up some work so the data inside the component is not updated until later :disappointed: When this happens we will be in a state were multiple tabs are active at once

Reply:
Yeah, I understand the problem. I'm just not quite sure I understand what this solution is actually doing. I find it incredibly hard to follow this short piece of code 😅

Would it be possible to refactor to something that might be a little easier to follow? If not, please add some explanatory comments, to give the rest of us a fighting chance 😄

Reply:
The solution saves the new state of the tabs in `this.tabs` so that we can be sure that it is up to date in case there is a new click before they are updated by the consumer.

I'm not sure how to refactor it, I was struggling a bit with the naming of the `mapTab` function. Maybe if we find a better name for that it will be more clear? :thinking: What part of the code do you think needs to be clarified? :slightly_smiling_face: 
We could also wrap line 185 above in another function with a name that perhaps could make it easier to understand

Reply:
If we can write some unit tests for `handleTabActivated`, I could take a look at refactoring it. Also, if we can't refactor it, tests would be really helpful for explaining how it works. It looks to me like we could break it out into a pure function pretty easily, and maybe put it in its own file and import it here?

Do you wanna, or should I? 🙂

Reply:
Yeah I can try and see if I can write some tests for it. I'm not sure how easy it will be making this a pure function though, since it both emits an event and saves state data :thinking: But maybe I'm misunderstanding something :smile:

Reply:
Ah, yeah, it emits an event too…

*(later…)*

I think this would work:

```ts
    private handleTabActivated(event: MDCTabBarActivatedEvent) {
        const results = this.inner(event);
        results.forEach(({tabs, changedTab}) => {
            this.changeTab.emit(changedTab);
            this.tabs = tabs;
        });
    }
    
    private inner(event: MDCTabBarActivatedEvent) {
        const index = event.detail.index;
        const oldSelectedTab = this.tabs.find((tab) => tab.active === true);
        let resultForOld: { tabs: Tab[]; changedTab: Tab; };

        if (oldSelectedTab) {
            resultForOld = this.replaceTab({ ...oldSelectedTab, active: false });
        }

        const resultForNew = this.replaceTab({ ...this.tabs[index], active: true });
        
        if (resultForOld) {
            return [resultForOld, resultForNew];
        }
        return [resultForNew];
        
    }

    private replaceTab(changedTab: Tab) {
        const tabs = this.tabs.map(this.mapTab(changedTab));
        return {tabs, changedTab};
    }

    private mapTab = (newTab: Tab) => (oldTab: Tab) => {
        if (newTab.id === oldTab.id) {
            return newTab;
        }

        return oldTab;
    };
```

Then the `inner` function should be easy to test.

Reply:
It's still accessing `this.tabs` though which makes it impure :slightly_frowning_face: 
I'm not sure if a unit test here would help with understanding the root cause of this issue, and I'm afraid that we make the code harder to understand if we refactor it around that :thinking: I've been trying to write an E2E test but it seems to be very flaky and Stencil does not seem to let me write the tests the way I want to :disappointed: I will have another go at it after lunch
I have at least renamed the functions a bit, so maybe it's easier to read now? Please let me know :slightly_smiling_face:

Reply:
> It's still accessing `this.tabs` though which makes it impure 🙁

Ah, crap, I missed that. But that should be easy to fix. Hang on!

Reply:
```ts
    private handleTabActivated(event: MDCTabBarActivatedEvent) {
        const results = this.inner(event, this.tabs);
        results.forEach(({tabs, changedTab}) => {
            this.changeTab.emit(changedTab);
            this.tabs = tabs;
        });
    }

    private inner(event: MDCTabBarActivatedEvent, incomingTabs: Tab[]) {
        const index = event.detail.index;
        const oldSelectedTab = incomingTabs.find((tab) => tab.active === true);
        let resultForOld: { tabs: Tab[]; changedTab: Tab; };

        if (oldSelectedTab) {
            resultForOld = this.replaceTab(incomingTabs, { ...oldSelectedTab, active: false });
        }

        incomingTabs = resultForOld?.tabs || incomingTabs;

        const resultForNew = this.replaceTab(incomingTabs, { ...incomingTabs[index], active: true });

        if (resultForOld) {
            return [resultForOld, resultForNew];
        }
        return [resultForNew];

    }

    private replaceTab(incomingTabs: Tab[], changedTab: Tab) {
        const tabs = incomingTabs.map(this.mapTab(changedTab));
        return {tabs, changedTab};
    }

    private mapTab = (newTab: Tab) => (oldTab: Tab) => {
        if (newTab.id === oldTab.id) {
            return newTab;
        }

        return oldTab;
    };
```

Reply:
Note, I'm not saying the above is better. It's just that it's testable. Once there are tests for it, I can refactor.

Reply:
Ok, I refactored it again and added both some E2E tests and some unit tests. I really don't feel that the tests illustrates the bug that this PR is solving though :smile: But at least there are some tests now! :partying_face: 
Please let me know if you think this is better or not :slightly_smiling_face:

Reply:
Great!

I pushed a new fixup with a small refactor of `tabs.ts`. It still does the same thing, I just made things a little more explicit, and added some comments.

👍 or 👎?

🙂